sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnInvokers.java

```diff
@@ -0,0 +1,500 @@
+      DoFnSignature signature, DoFn<InputT, OutputT> fn) {
+    Preconditions.checkArgument(
+        signature.getFnClass().equals(fn.getClass()),
+        "Signature is for class %s, but fn is of class %s",
+        signature.getFnClass(),
+        fn.getClass());
+    try {
+      @SuppressWarnings("unchecked")
+      DoFnInvoker<InputT, OutputT> invoker =
+          (DoFnInvoker<InputT, OutputT>)
+              getOrGenerateByteBuddyInvokerConstructor(signature).newInstance(fn);
+      return invoker;
+    } catch (InstantiationException
+        | IllegalAccessException
+        | IllegalArgumentException
+        | InvocationTargetException
+        | SecurityException e) {
+      throw new RuntimeException("Unable to bind invoker for " + fn.getClass(), e);
+    }
+  }
+
+  /**
+   * Returns a generated constructor for a {@link DoFnInvoker} for the given {@link DoFn} class and
+   * caches it.
+   */
+  private static synchronized Constructor<?> getOrGenerateByteBuddyInvokerConstructor(
+      DoFnSignature signature) {
+    Class<? extends DoFn> fnClass = signature.getFnClass();
+    Constructor<?> constructor = BYTE_BUDDY_INVOKER_CONSTRUCTOR_CACHE.get(fnClass);
+    if (constructor != null) {
+      return constructor;
+    }
+    Class<? extends DoFnInvoker<?, ?>> invokerClass = generateInvokerClass(signature);
+    try {
+      constructor = invokerClass.getConstructor(fnClass);
+    } catch (IllegalArgumentException | NoSuchMethodException | SecurityException e) {
+      throw new RuntimeException(e);
+    }
+    BYTE_BUDDY_INVOKER_CONSTRUCTOR_CACHE.put(fnClass, constructor);
+    return constructor;
+  }
+
+  /** Generates a {@link DoFnInvoker} class for the given {@link DoFnSignature}. */
+  private static Class<? extends DoFnInvoker<?, ?>> generateInvokerClass(DoFnSignature signature) {
+    Class<? extends DoFn> fnClass = signature.getFnClass();
+
+    final TypeDescription clazzDescription = new TypeDescription.ForLoadedType(fnClass);
+
+    DynamicType.Builder<?> builder =
+        new ByteBuddy()
+            // Create subclasses inside the target class, to have access to
+            // private and package-private bits
+            .with(
+                new NamingStrategy.SuffixingRandom("auxiliary") {
+                  @Override
+                  public String subclass(TypeDescription.Generic superClass) {
+                    return super.name(clazzDescription);
+                  }
+                })
+            // Create a subclass of DoFnInvoker
+            .subclass(DoFnInvoker.class, ConstructorStrategy.Default.NO_CONSTRUCTORS)
+            .defineField(
+                FN_DELEGATE_FIELD_NAME, fnClass, Visibility.PRIVATE, FieldManifestation.FINAL)
+            .defineConstructor(Visibility.PUBLIC)
+            .withParameter(fnClass)
+            .intercept(new InvokerConstructor())
+            // Delegate processElement(), startBundle() and finishBundle() to the fn.
```

Comment:
I'd propose removing this comment, or splitting it up (once for each delegation) since it'll be easier to maintain. Also update to include the invokeSetup and invokeTeardown methods.